Create IAM roles with operator

andy-paine opened the pull request:
- Pass in the permissions boundary ARN and the KIAM role ARN as
  variables
- Add terraform output and helm references to operator chart
- Operator is only allowed to modify/create roles that have the
  specified boundary which prevents it from having effectively full IAM
permissions

Commit:
Create IAM roles with operator

- Pass in the permissions boundary ARN and the KIAM role ARN as
  variables
- Add terraform output and helm references to operator chart
- Operator is only allowed to modify/create roles that have the
  specified boundary which prevents it from having effectively full IAM
permissions

blairboy362 commented on charts/gsp-cluster/templates/02-gsp-system/service-operator/crd.yaml:
```diff
@@ -93,6 +93,101 @@ status:
+  creationTimestamp: null
+  name: roles.iam.govsvc.uk
+spec:
+  group: iam.govsvc.uk
```
I guess we've abandoned our notion of abstracting away the underlying AWS types...

philandstuff commented on modules/gsp-cluster/service-operator.tf:
```diff
@@ -52,8 +52,24 @@ data "aws_iam_policy_document" "service-operator" {
+
+  statement {
+    actions = [
+      "iam:*",
```
```suggestion
      "iam:AttachRolePolicy",
      "iam:CreatePolicy",
      "iam:CreateRole",
      "iam:DeletePolicy",
      "iam:DeleteRole",
      "iam:DetachRolePolicy"
      "iam:PutRolePolicy"
      "iam:TagRole"
      "iam:UntagRole"
```
I worry about allowing `"iam:*"` because there's a lot of potential for privilege escalation. Not all IAM actions support the `iam:PermissionsBoundary` condition.  So, for example, someone could `iam:CreateAccessKey` on an existing User, even if that user were not subject to a permissions boundary (assuming such a user existed).  I'd rather be explicit about what we're allowing so I don't have to think about all these possible edge cases.

philandstuff commented on modules/gsp-cluster/service-operator.tf:
```diff
@@ -68,3 +84,24 @@ resource "aws_iam_policy_attachment" "service-operator" {
+      "rds-data:*",
+      "sqs:ReceiveMessage",
+      "sqs:DeleteMessage",
+      "sqs:GetQueueAttributes",
```
Missing `"sqs:SendMessage"`

Commit:
Ignore failing test

- This resource fails for an odd reason when creating even though it
  looks identical to the Postgres/SQS
- Apparently DCS are keen to kick the tyres on the broker so let's get
  this in an environment to see if it actually works

andy-paine commented on modules/gsp-cluster/service-operator.tf:
```diff
@@ -52,8 +52,24 @@ data "aws_iam_policy_document" "service-operator" {
     resources = [
-      "*" # TODO: revisit this
+      "*",
+    ] # TODO: revisit this
```
I'm not sure what this is about - I just ran the terraform format here

andy-paine replied:
I'm not sure what this is about - I just ran terraform format after making changes
